design/downstream-crl-design.md

```diff
@@ -0,0 +1,128 @@
+# Downstream Certificate Revocation List Proposal
+
+Status: Draft
+
+## Abstract
+This proposal covers the implementation of CRLs (Certificate Revocation List) in Contours DownstreamValidation, using the `crl` field in Envoy's [extensions.transport_sockets.tls.v3.CertificateValidationContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext) field.
+
+## Background
+
+Envoy supports specifying a PEM encoded CRL for use when validating a peer certificate.
+Currently Contour allows specifying a CA Certificate bundle that is used to validate a client.
+This is done using Envoy's Common TLS Context inside of a Downstream TLS context.
+Envoy's Common TLS Context also supports a CRL data source, which is a set of CRL's that are used when validating a clients certificate.
+This CRL data source can be exposed the same way that we expose the CA Certificate data source, through Contour's `DownstreamValidation` structure.
+This data source will give admins the ability to have finer grained control over their ingress, and enable CRLs for HTTPProxies that use client cert authentication
+
+## Goals
+- Allow a Kubernetes secret containing a PEM encoded CRL to be used when validating client connections
+- Perform simple validation of the K8s secret
+  - Validate that it is indeed a PEM CRL object
+
+## Non Goals
+- Upstream Validation
+- Support for `only_verify_leaf_cert_crl`
+  - Not in a released version of envoy yet, still in development
+
+
+## High-Level Design
+The design would for the most part mirror the implementation of `DownstreamValidation.CACertificate` and `DownstreamValidation.SkipClientCertValidation`.
+
+Sample YAML
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: client-validation-example
+  namespace: default
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: server-credentials
+      clientValidation:
+        caSecret: ca-cert-for-client-validation
+        caCrlSecret: crl-for-client-validation
```

Comment:
Could the name be just `crlSecret`? The CRLs are of course published by CAs but I don't recall bumping into CRL being referred as CA CRL.